### PR TITLE
Add functionality to make use of Config Classes in the container.

### PR DESCRIPTION
**Pull request by harikt**

### CODE REVIEW

**Comment by xtreamwayz** [comment]
I'm just wondering what it is you are trying to do here, since I'm not familiar with Aura.DI. I don't see how this would "make things faster" since all it does is adding more complexity. Some explanation would be helpful.

Reading the [link](http://auraphp.com/blog/2014/04/07/two-stage-config/) from the comments, this sounds like something to load configuration for modules or composer packages. And it is missing some essential parts like placing configuration in a ``Config`` class and adding modifications to ``composer.json`` to the ``extra->aura`` section.

**Comment by harikt** [comment]
Hi @xtreamwayz , 

Aura.Di provide a way to build it from config classes.  http://auraphp.com/packages/Aura.Di/config.html#1.8

Eg of a `Config` class :

```php
namespace Vendor\Package;

use Aura\Di\Container;
use Aura\Di\ContainerConfig;

class Config extends ContainerConfig
{
    public function define(Container $di)
    {
        $di->set('log_service', $di->lazyNew('Logger'));
        $di->params['Logger']['dir'] = '/path/to/logs';
    }

    public function modify(Container $di)
    {
        $log = $di->get('log_service');
        $log->debug('Finished config.');
    }
}
```
and you can build the container as

```
use Aura\Di\ContainerBuilder;

$container_builder = new ContainerBuilder();

// use the builder to create and configure a container
// using an array of ContainerConfig classes
$di = $container_builder->newConfiguredInstance([
    'Aura\Cli\_Config\Common',
    'Aura\Router\_Config\Common',
    'Aura\Web\_Config\Common',
]);
```

The problem here is the `$di` container will be locked and no more edit or adding settings is possible.

So these lines https://github.com/zendframework/zend-expressive-skeleton/blob/master/src/Composer/Resources/config/container-aura-di.php#L16-L24 

```php
// Inject config
$container->set('config', $config);
// Inject factories
foreach ($config['dependencies']['factories'] as $name => $object) {
    $container->set($object, $container->lazyNew($object));
    $container->set($name, $container->lazyGetCall($object, '__invoke', $container));
}
// Inject invokables
foreach ($config['dependencies']['invokables'] as $name => $object) {
    $container->set($name, $container->lazyNew($object));
}
```
will not work. 

And also if you look more closely what the `Config` builder is doing, you can see it need to set the values first.

And later it will try to get the values. In the current case as the configurations are not set https://github.com/zendframework/zend-expressive-skeleton/blob/master/src/Composer/Resources/config/container-aura-di.php#L16-L24  and it will break everything.

The code I have send in PR is exactly the same as from the `newConfiguredInstance` with moving this https://github.com/zendframework/zend-expressive-skeleton/blob/master/src/Composer/Resources/config/container-aura-di.php#L16-L24 piece of code inside it.

Let me know if I have explained it clearly. If you still have questions feel free to ask, so I can expand the particular portion.

Thank you.

**Comment by harikt** [comment]
By the way what I mean faster is creating the `Config` classes easier. I will edit the title.

File: src/Composer/Resources/config/container-aura-di.php [hunk 1]
```diff
@@ -9,6 +9,20 @@
 $builder = new ContainerBuilder();
 $container = $builder->newInstance();
 
+// Add config classes for two stage configuration
+// http://auraphp.com/blog/2014/04/07/two-stage-config/
+$configClasses = [
+    // 'Common',
```
Comment (harikt) [inline src/Composer/Resources/config/container-aura-di.php]: This is where the `Config` classes can be added.

**Comment by xtreamwayz** [comment]
I have tried this in relation to zend-expressive/issues#180. I begin to understand this now but I still don't see why someone would need this. I think it is easier to use proper Factories to inject dependencies. IMO it's also less overhead and faster in stead of trying to guess what to inject based on reflection. Besides that, the config in arrays can be cached and these Config classes can't be cached.

From their [blog](http://auraphp.com/blog/2014/04/07/two-stage-config/):

> With this two-stage configuration mechanism, using classes instead of includes, and using Composer to map the config modes the to the config classes, we have the benefits of both a define-only config system and a programmatic modification system. This gives us the best of both worlds, and helps us avoid resorting to static methods for configuration.

So basically this 2-step configuration is only there to prevent static configuration methods, which is something you wouldn't need anyway.

For the example you gave in zend-expressive/issues#180 you would get something like this:

```php
class HomePageFactory
{
    public function __invoke(ContainerInterface $container)
    {
        $template = $container->get(TemplateRendererInterface::class);
        $db = $container->get('database');

        return new HomePageAction($db, $template);
    }
}
```

**Comment by harikt** [comment]
@xtreamwayz when using Aura.Di I don't use factories. It is just another class which is not really needed when using Aura.Di .

**Comment by harikt** [comment]
@xtreamwayz by the way you can disregard the composer related functionalities. It was just to make things easier without adding classes to the configuration array ( https://github.com/zendframework/zend-expressive-skeleton/pull/14#discussion-diff-43838322R14 ) .

Thank you.

**Comment by weierophinney** [comment]
@harikt — While I "get" the idea behind the Aura.Di config classes, I see some fundamental issues that are hard to resolve with both current and proposed changes when used in the default Expressive configuration setup:

- Where should these classes exist? (i.e., should they be in the source tree? or the config tree? or somewhere entirely else?)
- Why the `_Config` convention? (I haven't seen anywhere in the Aura.Di docs indicating why, only that it's used, and I can't figure out how the autoloading should be defined.)
- How will it play with configuration caching?
- How will it play with configuration _merging_? (which is how *all* of the other containers allow for substitutions currently)

Can you provide some feedback towards these points?

**Comment by harikt** [comment]
Hi @weierophinney , 

> Where should these classes exist? (i.e., should they be in the source tree? or the config tree? or somewhere entirely else?)

These are just classes. It can exists any where and only thing is it need to be auto-loadable.

> Why the _Config convention? (I haven't seen anywhere in the Aura.Di docs indicating why, only that it's used, and I can't figure out how the autoloading should be defined.)

My assumption is you were looking at the examples 

```
'Aura\Cli\_Config\Common',
'Aura\Router\_Config\Common',
'Aura\Web\_Config\Common'
```

I have provided. It was a convention used in aura, but no one need to follow the same. Some examples are already provided in the docs http://auraphp.com/packages/Aura.Di/config.html#1.8 . Taking one  https://github.com/auraphp/Aura.Cli/blob/2.0.0/config/Common.php . The autoload is defined in [composer.json]( https://github.com/auraphp/Aura.Cli/blob/2.0.0/composer.json#L27 ) .

> How will it play with configuration caching?

If you are not using any closures the configuration can be cached. But it is not currently documented. ( https://github.com/auraphp/Aura.Di/blob/3.x/docs/serializing.md ) . I am not sure more on the caching of di and what are the plans in 3.0 . I think we should ping @pmjones to know more on this.

> How will it play with configuration merging? (which is how all of the other containers allow for substitutions currently)

I am not sure I get you here. Can you point me to the configuration merging? In my assumption the one thing containers were doing was looking whether a particular service is defined. If so get it and use it. Aura.Di provides the service and it depends on how the configurations are used.

**Comment by weierophinney** [comment]
Closing. We currently pin to Aura.Di 3.0, and I'm not satisfied that this will be immediately understood by developers.
